play-framework/async-http/test/controllers/HomeControllerTest.java

```diff
@@ -0,0 +1,184 @@
+package controllers;
+
+import java.util.concurrent.ExecutionException;
+import org.junit.Before;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import play.Application;
+import play.inject.guice.GuiceApplicationBuilder;
+import play.libs.concurrent.Futures;
+import play.libs.ws.WSClient;
+import play.libs.ws.WSResponse;
+import play.libs.ws.ahc.AhcCurlRequestLogger;
+import play.mvc.Result;
+import play.mvc.Results;
+import play.test.WithServer;
+
+import java.time.Duration;
+import java.util.OptionalInt;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+import static java.time.temporal.ChronoUnit.SECONDS;
+import static org.junit.Assert.assertEquals;
+import static play.mvc.Http.Status.SERVICE_UNAVAILABLE;
+
+import play.mvc.Http.MultipartFormData.*;
+import akka.stream.javadsl.*;
+import akka.util.ByteString;
+
+public class HomeControllerTest extends WithServer {
+
+    private final Logger log = LoggerFactory.getLogger(HomeControllerTest.class);
+    private String url;
+    private int port;
+
+    @Override
+    protected Application provideApplication() {
+        return new GuiceApplicationBuilder().build();
+    }
+
+    @Before
+    public void setup() {
+        OptionalInt optHttpsPort = testServer.getRunningHttpsPort();
+        if (optHttpsPort.isPresent()) {
+            port = optHttpsPort.getAsInt();
+            url = "https://localhost:" + port;
+        } else {
+            port = testServer.getRunningHttpPort().getAsInt();
+            url = "http://localhost:" + port;
+        }
+    }
+
+    @Test
+    public void givenASingleGetRequestWhenResponseThenBlockWithCompletableAndLog()
+      throws InterruptedException, ExecutionException {
+        AtomicInteger completedReqs = new AtomicInteger(0);
+        WSClient ws = play.test.WSTestClient.newClient(port);
+        WSResponse wsResponse = ws.url(url)
+                                  .setRequestFilter(new AhcCurlRequestLogger())
+                                  .addHeader("key", "value")
+                                  .addQueryParameter("num", "" + 1)
+                                  .get()
+                                  .toCompletableFuture()
+                                  .get();
+
+        log.debug("Thread#" + Thread.currentThread()
+                                    .getId() + " Request complete: Response code = "
+          + wsResponse.getStatus()
+          + " | Response: " + wsResponse.getBody() + " | Current Time:"
+          + System.currentTimeMillis());
+    }
+
+    @Test
+    public void givenASingleGetRequestWhenResponseThenLog() throws InterruptedException {
+        AtomicInteger completedReqs = new AtomicInteger(0);
```

Comment:
You could replace this with a `CountDownLatch` which would allow the thing at the end to wait for the latch to get to 0. That might be cleaner than the while loop with the sleep.

Reply:
done